File: update-launchers.go
package main

import (
	"fmt"
	"io"
	"io/ioutil"
	"log"
	"os"
	"os/exec"
	"os/user"
	"strings"
)

type launcher struct {
	name      string
	path      string
	filename  string
	deb       string
	sandboxed bool
	todelete  bool
	toupdate  bool
}

func main() {
	if uid, _ := user.Current(); uid.Uid != "0" {
		os.Stderr.WriteString(fmt.Sprintf("Wrong user %s! - Run this program as root\n", uid.Name))
		os.Exit(1)
	}

	fmt.Println("Scanning application launchers")

	// get list of files in /usr/share/parrot-menu/applications (source)""
	sfiles, err := ioutil.ReadDir("/usr/share/parrot-menu/applications/")
	if err != nil {
		log.Fatal(err)
	}

	// get list of files in /usr/share/applications (destination)
	dfiles, err := ioutil.ReadDir("/usr/share/applications/")
	if err != nil {
		log.Fatal(err)
	}

	// get list of installed packages
	tmp, _ := exec.Command("bash", "-c", "apt list --installed | cut -d'/' -f1").Output()
	packages := strings.Split(string(tmp), "\n")

	// create source and destination structures
	source := make([]launcher, len(sfiles))
	destination := make([]launcher, len(dfiles))

	// initialize source structure
	var spath string
	for i, f := range sfiles {
		spath = fmt.Sprintf("/usr/share/parrot-menu/applications/%s", f.Name())
		getLauncherContent(&source[i], spath, f.Name())
	}

	// initialize destination structure
	var dpath string
	for i, f := range dfiles {
		dpath = fmt.Sprintf("/usr/share/applications/%s", f.Name())
		getLauncherContent(&destination[i], dpath, f.Name())
	}

	fmt.Println("Updating active launchers")

	removeOldLaunchers(destination)
	copyActiveLaunchers(source, packages)
	exec.Command("sync")

	fmt.Println("Done")
}

func getLauncherContent(s *launcher, path string, filename string) {
	// set path
	s.path = path

	// set filename
	s.filename = filename

	// set name
	tmpname, _ := exec.Command("bash", "-c",
		fmt.Sprintf("grep Name= %s | sed -e 's/Name=//g'", path)).Output()
	s.name = strings.Trim(string(tmpname), "\n")

	// set package name
	tmpdeb, _ := exec.Command("bash", "-c",
		fmt.Sprintf("grep X-Parrot-Package= %s | sed -e 's/X-Parrot-Package=//g'", path)).Output()
	s.deb = strings.Trim(string(tmpdeb), "\n")
}

func removeOldLaunchers(files []launcher) {
	for _, f := range files {
		if f.deb != "" {
			os.Remove(f.path)
		}
	}
}

func copyLauncher(source string, destination string) (error, error, error) {
	in, errOpen := os.Open(source)
	if errOpen != nil {
		fmt.Printf("[WARNING] Can't open: %s - %v\n", source, errOpen)
	}

	out, errClose := os.Create(destination)
	if errClose != nil {
		fmt.Printf("[WARNING] Can't create: %s - %v\n", destination, errClose)
	}

	_, errCopy := io.Copy(out, in)
	if errCopy != nil {
		fmt.Printf("[WARNING] Can't install launcher - %v\n", errCopy)
	}

	defer in.Close()
	defer out.Close()
	return errOpen, errClose, errCopy
}

func copyActiveLaunchers(launchers []launcher, packages []string) {
	for _, l := range launchers {
		for _, p := range packages {
			if l.deb == p {
				copyLauncher(l.path, fmt.Sprintf("/usr/share/applications/%s", l.filename))
			}
		}
	}
}

File: update_launchers.nim
# Update the lauchers in /usr/share/parrot-menu/applications
# Copy launchers from parrot-menu (/usr/share/parrot-menu/applications) to dest folder if it was installed
#   1. Check package name from X-Parrot-Package=
#   2. Check if it is installed using apt list
#   3.
#     a. If it is not in the list -> Remove it in destination folder
#     b. If it is in the list -> copy it to dest folder if it doesnt exists
#   4. Fix duplicate launchers
# Remove old launchers that removed or uninstalled

import os, strutils

const
  dirLauncherSource = "/usr/share/parrot-menu/applications/"
  dirLaucherDest = "/usr/share/applications/"
  path = "/var/lib/dpkg/status"


proc getXPackageName(path: string): string =
  for line in lines(path):
    # Normaly it starts with X-Parrot-Packages
    # But some launchers might contains X-Parrot-package (typo mistake)
    # So we use keyword X-Parrot only
    if line.startsWith("X-Parrot-"):
      return line.split("=")[1]
  return ""


proc query_installed(): seq[string] =
  var
    pkg_name: string
  for line in lines(path):
    if line.startsWith("Package: "):
      pkg_name = line.split(": ")[1]
    elif line.startsWith("Status: "):
      if line == "Status: install ok installed":
        result.add(pkg_name)


proc fixDebLaunchers() =
  #[
    There are packages from Debian that has custom launchers
    It makes error after install pentest tools from Home Edition
    or create duplicate launchers in menu.
    We are removing them here
    chirp will be removed because of python2 removal problem
  ]#
  let blacklistLauncherName = [
    "org.radare.Cutter.desktop",
    "gpa.desktop",
    "rtlsdr-scanner.desktop",
    "gnuradio-grc.desktop",
    "arduino.desktop",
    "gqrx.desktop",
    "zulucrypt-gui.desktop",
    "zulumount-gui.desktop",
    "ophcrack.desktop",
    "xsser.desktop",
    "io.github.mhogomchungu.sirikali.desktop",
    "etherape.desktop",
    "edb.desktop",
    "lynis.desktop",
    "wireshark.desktop",
    "ettercap.desktop",
    "chirp.desktop",
    "hydra-gtk.desktop",
    "driftnet.desktop",
    "gscriptor.desktop",
    "spectool_gtk.desktop",
    "gksu.desktop",
    "re.rizin.cutter.desktop", # Duplicate launcher of rizin's cutter
    "openjdk-8-policytool.desktop",
  ]
  for fileName in blacklistLauncherName:
    let finalPath = dirLaucherDest & fileName
    if fileExists(finalPath):
      if not tryRemoveFile(finalPath):
        stderr.write("[x] Error while removing " & finalPath & "\n")


proc fixOldLaunchers(path: string) =
  #[
    Some new launchers have been changed name to parrot-* to serv-*
    We try to remove duplicated launchers
  ]#
  let
    fileName = splitPath(path).tail
    newNameArr = ["serv-"]
  var
    destName: string = dirLaucherDest
    isDeleteNeeded: bool

  for checkName in newNameArr:
    isDeleteNeeded = false
    if fileName.startsWith(checkName):
      isDeleteNeeded = true
      destName &= "parrot-" & fileName.replace(checkName, "")
      break

  if isDeleteNeeded:
    if not tryRemoveFile(destName):
      stderr.write("[x] Error while removing " & destName & "\n")


proc update_launchers() =
  # Get all installed packages
  let installed = query_installed()

  # Get all file in applications
  var allLaunchers: seq[string]
  for kind, path in walkDir(dirLauncherSource):
    let fileName = splitPath(path).tail
    allLaunchers.add(fileName)
    if fileName.startsWith("parrot-") or fileName.startsWith("serv-"):
      let
        aptParrotPackage = getXPackageName(path)
      if aptParrotPackage != "":
        #[
          1. Case 1: if the package is installed but
            a) Add launcher if it isn't there
            b) Compare launcher's data and update it
          2. Case 2: if the package isn't installed, remove launcher
        ]#

        try:
          let finalDestPath = dirLaucherDest & fileName
          # If the package is installed
          # Check if package name is in installed list. The name matches a line exactly
          # if contains(installed, re("(^|\\n)" & aptParrotPackage & "($|\\n)")):
          if installed.contains(aptParrotPackage):
            # Check if file is not in the final directory
            if not fileExists(finalDestPath) or not sameFileContent(path, finalDestPath):
              # Update new launcher
              try:
                # If file does not exists in dest folder, copy it
                copyFile(path, finalDestPath)
              except:
                stderr.write("[x] Error while copying file " & path & " to " & finalDestPath & "\n")
          else:
            if fileExists(finalDestPath):
              # Remove old launchers here
              if not tryRemoveFile(finalDestPath):
                stderr.write("[x] Error while processing " & path & "\n")
          
          # In this version, we are moving name to serv-, native and being more with different categories
          fixOldLaunchers(path)
        except:
          stderr.write("[x] Error while processing " & path & "\n")
          echo getCurrentExceptionMsg()

  for kind, path in walkDir(dirLaucherDest):
    let currentLauncher = splitPath(path).tail
    # Check if the launcher is Parrot's specific
    if (currentLauncher.startsWith("parrot-") or currentLauncher.startsWith("serv-")) and currentLauncher.endsWith(".desktop"):
      # Get package name from launcher. If package name != "" then it belongs to parrot-menu (or old one)
      # We can use the fileExist from source method because some packages are having custom launcher in the package
      let packageOfLauncher = getXPackageName(path)
      if packageOfLauncher != "":
        # We must test if package is not installed here
        if not allLaunchers.contains(currentLauncher) and not installed.contains(packageOfLauncher):
          if not tryRemoveFile(path):
            stderr.write("[x] Error while removing " & path & "\n")

echo "Scanning application launchers"
update_launchers()
echo "Removing duplicate launchers or broken launchers"
fixDebLaunchers()
echo "Launchers are updated"
